drivers/phoenixcontact_modbus.c

```diff
@@ -49,21 +54,58 @@ void upsdrv_initinfo(void)
 	upsdebugx(2, "upsdrv_initinfo");
 
 	dstate_setinfo("device.mfr", "Phoenix Contact");
-	dstate_setinfo("device.model", "QUINT-UPS/24DC");
-
+	
 	/* upsh.instcmd = instcmd; */
 	/* upsh.setvar = setvar; */
+
+	uint16_t FWVersion;
+
+	mrir(modbus_ctx, 0x0004, 1, &FWVersion);
+
+	if (FWVersion == 544)
+	{
+		UPSModel = OLD_UPS;
+		dstate_setinfo("device.model", "QUINT-UPS/24DC OLD");
+	}
+	else if (FWVersion == 305)
+	{
+		UPSModel = NEW_UPS;
+		dstate_setinfo("device.model", "QUINT-UPS/24DC NEW");
+	}
 }
 
 void upsdrv_updateinfo(void)
 {
 	uint16_t tab_reg[64];
-
+	
 	errcount = 0;
 
 	upsdebugx(2, "upsdrv_updateinfo");
 
-	mrir(modbus_ctx, 29697, 3, tab_reg);
+	uint16_t actual_code_functions;
+	uint16_t actual_alarms = 0;
+	uint16_t actual_alarms1 = 0;
+
+	switch (UPSModel)
+	{
+	case NEW_UPS:
+		
+		mrir(modbus_ctx, 0x2000, 1, &actual_code_functions);
+
+		tab_reg[0] = CHECK_BIT(actual_code_functions, 2); //battery mode is the 2nd bit of the register 0x2000
+		tab_reg[2] = CHECK_BIT(actual_code_functions, 5); //battery charging is the 5th bit of the register 0x2000
+				
+		mrir(modbus_ctx, 0x3001, 1, &actual_alarms1);
+
+		tab_reg[1] = CHECK_BIT(actual_alarms1, 2); //battery discharged is the 2nd bit of the register 0x3001
+		break;
+	case OLD_UPS:
+
+		mrir(modbus_ctx, 29697, 3, tab_reg); /* LB is actually called "shutdown event" on this ups */
```

Comment:
Any idea if these numbers make more sense in hex? Are they perhaps bit-masks of some command combos?

Reply:
No, got no particular preference here. If maintenance would be easier with decimal numbers - so be it. Maybe worth a short comment at start of the file though.